Allow cross-repository dependencies on issues

bhalbright opened the pull request:
Currently issues can be added as dependencies within the same repository. This changes so that issues in other repositories can be added as dependencies. The current issue search only operates in the same repository, so I added a new api method to search for issues across repositories (the repositories the user can access - the repositories searched would be the same repositories that the user sees on their dashboard).

Ref issue #7405

guillep2k commented on models/issue.go:
```diff
@@ -1834,7 +1834,7 @@ type DependencyInfo struct {
 func (issue *Issue) getBlockedByDependencies(e Engine) (issueDeps []*DependencyInfo, err error) {
 	err = e.Table("issue_dependency").
-		Select("issue.id, issue.repo_id, issue.index, issue.is_closed, issue.name").
+		Select("issue.*").
```
Just asking... is it necessary to select any columns from `Issue` to build a `DependencyInfo`?

bhalbright replied:
Thanks @guillep2k for taking a look at this. I "think" it is necessary to pull info from the issues table to properly get the relevant dependency info (i.e. that is needed for display purposes). Originally I was only selecting the necessary columns, but I changed to issue.* because it was breaking the build in mysql and sqlite (though I'm not sure why, it was working locally with mysql). I'm totally open for suggestions thought. I'm new to this project and go so I could definitely be missing the optimal solution. Thanks again!

guillep2k replied:
I think it's like a shortcut you're taking, there. You're forcing xorm to bring all the related issues for displaying, but perhaps it's not the most expected place to do that. The fact that you've been asking for some data is what might have confused the ORM. I'm no expert on it either. Have you tried selecting no columns at all from `Issues`? I can't check ATM.

bhalbright replied:
I tried selecting no columns and it worked. I also tried just leaving the original code which was 'Select("issue.*")', so I'll just leave that unless you feel I should remove it, thanks

guillep2k replied:
Oh, I see, you did not introduce that line. It seems to be just unnecesary code; perhaps an attempt to bring the rows to the ORM ahead of time, but I made some tests with and without that line and it doesn't change the number of queries made to the database. So, I feel that this is what we should be doing:
```
	err = e.Table("issue_dependency").
		Join("INNER", "issue", "issue.id = issue_dependency.dependency_id").
		Where("issue_id = ?", issue.ID).
		Find(&issueDeps)
```

bhalbright replied:
thanks @guillep2k I've updated it

bhalbright replied:
well haha actually that broke it, the id in my struct is being filling with the id from the issue_dependency table instead of the id from the issue table that I need. From taking a look at the xorm join docs, the returned columns need to be in the right order, so I probably need to re-structure the queries, or just use a straight sql in xorm, but straight sql seems to be used very sparingly in gitea so I should probably do the former.

guillep2k replied:
If ever possible, direct SQL should be the last resort. :grin:

bhalbright replied:
I changed it to a more xorm friendly query I believe. for example: 
`	err = e.Table("issue").
		Join("INNER", "repository", "repository.id = issue.repo_id").
		Join("INNER", "issue_dependency", "issue_dependency.dependency_id = issue.id").
		Where("issue_id = ?", issue.ID).
		Find(&issueDeps)
`

lunny commented on options/locale/locale_cs-CZ.ini:
```diff
@@ -908,7 +908,6 @@ issues.dependency.add_error_dep_issue_not_exist=Související úkol neexistuje.
 issues.dependency.add_error_dep_not_exist=Závislost neexistuje.
```
You MUST not change the translation files directly but go to https://crowdin.com/project/gitea

bhalbright replied:
sorry...I will revert these...in this case I was removing an error message that is no longer used with this change. if there's anything special to this case let me know or I'll just read on crowdin to see what I need to do thanks

guillep2k commented:
It's a nice work. I sill haven't had time to look into the code, but here's my impressions using it.

![image](https://user-images.githubusercontent.com/18600385/65572061-36637e80-df3d-11e9-85de-5ff55d41178d.png)

The "salmon color" cross shows up "floating" in an eerie fashion. Perhaps right-aligned with the number?

Also:
![image](https://user-images.githubusercontent.com/18600385/65572139-7aef1a00-df3d-11e9-9099-70a26c016fd5.png)

Is the + symbol something expected to mean reference? Perhaps 
![image](https://user-images.githubusercontent.com/18600385/65572666-29e02580-df3f-11e9-8f09-4b2293ba37e3.png)  or
![image](https://user-images.githubusercontent.com/18600385/65572685-382e4180-df3f-11e9-83cb-57e2e362680d.png) would work better?

guillep2k commented:
![image](https://user-images.githubusercontent.com/18600385/65572854-b985d400-df3f-11e9-965b-7b796a5535bf.png)

Perhaps add a visual cue when it is no longer blocking (**it's already closed**).

guillep2k commented:
Otherwise, great work!!

guillep2k commented:
OK, I've tested on my site and this is what I've observed:

* _The issue selection pop-up lists the issues in some strange order_

![image](https://user-images.githubusercontent.com/18600385/65609639-7e5ac380-df86-11e9-9325-72bd0be3262b.png)

I think perhaps the best order to display is:
* First, all the opened issues from the current repository, ordered by index (descending).
* Next, all the opened issues from other repositories, ordered by creation date (descending).

Of course, with a proper count limit.

* _Once an issue is selected, it cannot be unselected (I'm forced to add it and then delete it)_

![image](https://user-images.githubusercontent.com/18600385/65610566-ebbb2400-df87-11e9-9a63-b99ac7470a50.png)

There should be some mechanism to unselect the issue from the box.

Otherwise, this is a great job. I like it a lot!

guillep2k commented on models/issue.go:
```diff
@@ -1790,8 +1798,8 @@ func GetRepoIssueStats(repoID, uid int64, filterMode int, isPull bool) (numOpen
-	var repoCond = builder.Eq{"repo_id": repoID}
+func SearchIssueIDsByKeyword(kw string, repoIDs []int64, limit, start int) (int64, []int64, error) {
+	var repoCond = builder.In("repo_id", repoIDs)
 	var subQuery = builder.Select("id").From("issue").Where(repoCond)
```
Perhaps you should just search in the first repo (usually the local repo) and only if limit is not reached, then search in the other repos. In either case you should at least sort locally by ID descending:
```
import "sort"
sort.Slice(ids, func(i, j int) bool {return ids[i] > ids[j]})
```

bhalbright replied:
this makes sense when using the db option for issue indexing, but if using bleve, I'm guessing I should leave that as is, since by default my understanding is that the sort is by relevancy scoring, but I could be corrected. And I think this code is only hit when a keyword is actually typed, otherwise it is just a list of issues up to the limit (i.e. not running this query) so I'll visit that code too and see what the best way to get a predictable listing is.

guillep2k replied:
I see. I guess there's not much to do about the order returned by Bleve. However, it is possible to prioritize the local repository because Bleve keeps separate indexes for each one and the search is done one by one. I didn't check, perhaps it comes out naturally?

bhalbright replied:
I looked at it a little more tonight and I think I have a better plan to make it straightforward I hope. I feel a little silly but I believe the ordering is actually coming out of a call to models.Issues where we pass the issue ids after we've found the list, so the sorting in these search methods isn't important. I think I can apply the proper sorting there so it would be consistent for db or bleve. Hopefully in a few days I'll be able to commit some changes that hopefully resolve the issues you pointed out related to this thanks!

bhalbright replied:
I believe I fixed the sorting of the issues the way you proposed but it isn't done here.

guillep2k commented on models/issue.go:
```diff
@@ -1880,33 +1888,39 @@ func UpdateIssueDeadline(issue *Issue, deadlineUnix timeutil.TimeStamp, doer *Us
+// DependencyInfo represents high level information about an issue which is a dependency of another issue.
+type DependencyInfo struct {
+	Issue      `xorm:"extends"`
+	Repository `xorm:"extends"`
```
Nice! I didn't know this could be done.

guillep2k commented on models/issue.go:
```diff
@@ -1880,33 +1888,39 @@ func UpdateIssueDeadline(issue *Issue, deadlineUnix timeutil.TimeStamp, doer *Us
+		Join("INNER", "repository", "repository.id = issue.repo_id").
+		Join("INNER", "issue_dependency", "issue_dependency.dependency_id = issue.id").
 		Where("issue_id = ?", issue.ID).
 		Find(&issueDeps)
```
Some sort criteria should be used here, too.

bhalbright replied:
I added criteria to sort by the repo and issue id

bhalbright replied:
done, thanks for the case statement suggestion

guillep2k commented on models/issue.go:
```diff
@@ -1880,33 +1888,39 @@ func UpdateIssueDeadline(issue *Issue, deadlineUnix timeutil.TimeStamp, doer *Us
+		Join("INNER", "repository", "repository.id = issue.repo_id").
+		Join("INNER", "issue_dependency", "issue_dependency.issue_id = issue.id").
 		Where("dependency_id = ?", issue.ID).
 		Find(&issueDeps)
```
And here.

bhalbright replied:
I also added criteria to sort by the repo and issue id

bhalbright replied:
done

guillep2k commented on routers/api/v1/repo/issue.go:
```diff
@@ -22,6 +23,123 @@ import (
+		issueIDs, err = issue_indexer.SearchIssuesByKeyword(repoIDs, keyword)
+	}
+
+	if splitted := strings.Split(ctx.Query("labels"), ","); len(splitted) > 0 {
```
This condition will always be true. `strings.Split()` will return `nil` only if both `s` and `sep` are empty.
https://golang.org/pkg/strings/#Split

bhalbright replied:
I'm not sure why it would always return true as I was thinking the condition is based on the length of the slice produced from the split, but I broke it up to make it more clear in case I'm not understanding.

guillep2k commented on routers/api/v1/repo/issue.go:
```diff
@@ -22,6 +23,123 @@ import (
 )
 
+// SearchIssues searches for issues across the repositories that the user has access to
+func SearchIssues(ctx *context.APIContext) {
```
@lafriks  would you care to take a look at this API?

bhalbright commented:
> It's a nice work. I sill haven't had time to look into the code, but here's my impressions using it.
> 
> ![image](https://user-images.githubusercontent.com/18600385/65572061-36637e80-df3d-11e9-85de-5ff55d41178d.png)
> 
> The "salmon color" cross shows up "floating" in an eerie fashion. Perhaps right-aligned with the number?
> 
> Also:
> ![image](https://user-images.githubusercontent.com/18600385/65572139-7aef1a00-df3d-11e9-9099-70a26c016fd5.png)
> 
> Is the + symbol something expected to mean reference? Perhaps
> ![image](https://user-images.githubusercontent.com/18600385/65572666-29e02580-df3f-11e9-8f09-4b2293ba37e3.png) or
> ![image](https://user-images.githubusercontent.com/18600385/65572685-382e4180-df3f-11e9-83cb-57e2e362680d.png) would work better?

Thanks, I'll fix the alignment of the remove icon. On the "+" icon, that wasn't something I changed, but I'm guessing it means that the dependency was added. Would you still like me to change it?

bhalbright commented:
> ![image](https://user-images.githubusercontent.com/18600385/65572854-b985d400-df3f-11e9-965b-7b796a5535bf.png)
> 
> Perhaps add a visual cue when it is no longer blocking (**it's already closed**).

Interesting, I'll try to reproduce this, it is supposed to look like this (this was actually existing logic I didn't change):

![image](https://user-images.githubusercontent.com/32200834/65655694-a8da6a00-dfe2-11e9-8818-7271f00c18c3.png)

guillep2k commented:
> Thanks, I'll fix the alignment of the remove icon. On the "+" icon, that wasn't something I changed, but I'm guessing it means that the dependency was added. Would you still like me to change it?

No, sorry, I didn't realize it was there before. It's a subject for another PR one day, then.

guillep2k commented:
> 
> 
> > ![image](https://user-images.githubusercontent.com/18600385/65572854-b985d400-df3f-11e9-965b-7b796a5535bf.png)
> > Perhaps add a visual cue when it is no longer blocking (**it's already closed**).
> 
> Interesting, I'll try to reproduce this, it is supposed to look like this (this was actually existing logic I didn't change):
> 
> ![image](https://user-images.githubusercontent.com/32200834/65655694-a8da6a00-dfe2-11e9-8818-7271f00c18c3.png)

I'll check again. I might have failed to copy the templates properly.

bhalbright commented:
> 
> No, sorry, I didn't realize it was there before. It's a subject for another PR one day, then.

Thanks! And thanks for your kind words and detailed review, I'll work through everything you pointed out.

guillep2k commented:
I can confirm that the missing strike-through was a problem on my side. It works as it should:

![image](https://user-images.githubusercontent.com/18600385/65656157-39ba4100-dff5-11e9-8392-385a322bfc64.png)

Interestingly (and this -I think- is also **something for another PR**), this is confusing:

![image](https://user-images.githubusercontent.com/18600385/65656204-5c4c5a00-dff5-11e9-9eb1-f0290226b5b6.png)

The above picture is from a _closed_ PR which is no longer blocking the dependent issue. I think it should be displayed as strike-through as well. The situation is the same whether it's _closed_ or _merged_. And here's the interesting part: _should an issue that depends on a closed (unmerged) PR be considered unblocked?_

guillep2k commented on public/js/index.js:
```diff
@@ -3096,11 +3096,10 @@ function deleteDependencyModal(id, type) {
         .dropdown({
             apiSettings: {
-                url: suburl + '/api/v1/repos/' + repolink + '/issues?q={query}',
+                url: suburl + '/api/v1/repos/issues/search?q={query}',
```
I think you should pass the repo id or name, and prioritize issues from that repository. There can be thousands of issues in a Gitea installation (Gitea alone has passed 8200).

bhalbright replied:
thanks, this is done now

guillep2k commented:
@bhalbright Perhaps it's more important to concentrate in "correctness" for the moment, and leave the result order for another "improvement" PR. Check my comments about the API.

bhalbright commented:
> @bhalbright Perhaps it's more important to concentrate in "correctness" for the moment, and leave the result order for another "improvement" PR. Check my comments about the API.

@guillep2k good point, I'll focus on the issues first.

bhalbright commented on models/issue.go:
```diff
@@ -377,6 +377,14 @@ func (issue *Issue) apiFormat(e Engine) *api.Issue {
 		Updated:  issue.UpdatedUnix.AsTime(),
 	}
 
+	if issue.Repo != nil {
```
I agree, there's a call to loadRepo prior to this block so it is unnecessary. I removed the check

guillep2k commented on models/issue.go:
```diff
@@ -1901,6 +1899,7 @@ func (issue *Issue) getBlockedByDependencies(e Engine) (issueDeps []*DependencyI
 		Join("INNER", "repository", "repository.id = issue.repo_id").
 		Join("INNER", "issue_dependency", "issue_dependency.dependency_id = issue.id").
 		Where("issue_id = ?", issue.ID).
+		OrderBy("repository.id, issue.id ASC").
```
Mmm.... If you sort by issue.id DESC you get the newer issues first, which would make sense. Order by repository ID makes no sense. It's an internal number.
Perhaps:
```
OrderBy("CASE WHEN repository.id = ? THEN 1 ELSE 2 END, issue.id DESC", current-repo-id)
```
would cause the issues from the current repo to show first.

bhalbright replied:
gotcha, these are for the list of blocked by or blocking dependencies displayed on the issue page (i.e. not the drop-down list to add a new one). I was thinking it would make sense for the issues to be in order of their id grouped by repo, so you'd see all the "repo1" issues and then "repo2" issues, not mixed together. But I agree with your suggestion, thanks!

guillep2k replied:
Oh, I see! To group by repo and still have the current repo top of the list:
```
OrderBy("CASE WHEN repository.id = ? THEN 0 ELSE repository.id END, issue.id DESC", current-repo-id)
```
But I've got confused. I thought this was for the search function.

bhalbright replied:
thanks for the suggestion, I used the order by statement you suggested

bhalbright commented on routers/api/v1/repo/issue.go:
```diff
@@ -22,6 +23,123 @@ import (
+	repoIDs := make([]int64, 0)
+	issueCount := 0
+	for page := 1; ; page++ {
+		repos, count, err := models.SearchRepositoryByName(&models.SearchRepoOptions{
```
this is finding the repository ids that the user has access to, then this list passed later in the method to search for issues across the repositories, the method to find the repositories uses paging, so this code runs through the pages until it gets all of them. There's some very similar code to this in a couple other places

bhalbright commented on routers/api/v1/repo/issue.go:
```diff
@@ -22,6 +23,123 @@ import (
+			Collaborate: util.OptionalBoolNone,
+			UserIsAdmin: ctx.IsUserSiteAdmin(),
+			UserID:      ctx.Data["SignedUserID"].(int64),
+			OrderBy:     models.SearchOrderByID,
```
I'm not sure I understand...do you mean we shouldn't search for issues across all the repositories the user has access to, but rather search across the let's say 15 most recently updated repositories that the user has access to? I could imagine some users thinking this is a bug, like I want to add a dependency to an issue in repository X, but the search won't list issues from the repository X (because the user has access to more than 15 repositories and repository X hasn't been updated in a while). Sorry if I'm just missing the point, thanks.

guillep2k replied:
We need to limit somehow the burden on the database for actions like this. Think of a bigger gitea installation, with hundreds of public repositories. From the user's point of view, it's also useless to have 100's of results and the one they want is on page 34. If we have to limit them in some way (and we must), chances are that dependencies are from the most active repositories.

There are other ways to address this problem, but they will demand more work. I'm just trying to find the shortest path to get this PR added to the main code. As I've said, there's always the possibility of a future PR with more enhancements. For example, we could let the user type in the issue in a text/list combo box, instead of using a list box. Other possibility is to have 15 as a default and let the admin change the value in `app.ini`.

jag3773 replied:
As a user, if I'm searching for an issue that I know is out there, I want it to show up, regardless of age.  I'm in favor of prioritizing the list by recently active but I think only searching on the recent issues would be a mistake.

If I search for the exact title of an issue in the 16th most active repository, what happens?

Also, how do we *know* that searching 10,000 issues will cause performance issues?

I could certainly live with an `app.ini` setting as we could increase it to a high number.  However, just noting that I think the whole premise will yield unexpected results for a user.

bhalbright replied:
thanks @guillep2k for your thoughts. I like the idea of using an app.ini setting to default the setting but allows users who want a deeper search to increase the value.

guillep2k replied:
@bhalbright if you'd like to go down that road, I'd suggest a setting that is the max number of repositories, 0 for none (it doesn't cost much to do it) and -1 for all of them. In any way, I'd use a lowish value as default (10~15).

bhalbright replied:
Hi, @guillep2k I wanted to run a different idea by you to see if you like it. Rather than implementing a setting in app.ini for a "repository search depth", how about instead implementing a setting in app.ini to enable "cross-repository dependencies". If enabled, when you search for issues to add as a dependency, it will search across all the repositories the user is granted access to (the way I have it implemented now). But if turned off, the search will work as it is today (search only the current repository). IMO this will be a little easier to implement. If a larger gitea installation isn't happy with the performance, they can just turn the feature off.

bhalbright replied:
I went ahead for now and implemented like I mentioned (either having cross-repo dependencies on or off), hope you don't mind. It seems to be working, but I'm going to test some more before I push up the commits

bhalbright replied:
I've commited these changes now (using a setting to turn on/off cross repo dependencies)

bhalbright commented:
@guillep2k in my latest commits I fixed the issue with the "delete" button not lining up with the issue number, i.e.  this one:
![image](https://user-images.githubusercontent.com/32200834/66263910-a802be80-e7c0-11e9-8af5-c92c9e3b81d6.png)

I see your point about not being able to clear the result in the search box. I did a little looking and I see Gitea is using the Semantic UI library there, specifically version 2.3.1. It looks like in the v2.4 they added a "clearable" property you can add to the dropdown to do what you had in mind. So I'm not sure what to do for now, I'm sure I could come up with something on my own, but maybe it would be better to do a PR in the future if/when the UI components are updated.

bhalbright commented:
@guillep2k  I think I've addressed the issues you brought up in the commits I have added. I'm honestly not used to the github interface for a PR with a bunch of comments and stuff so I hope I'm doing this right :) Let you know what you think at your convenience.

guillep2k commented on routers/api/v1/repo/issue.go:
```diff
@@ -105,7 +110,8 @@ func SearchIssues(ctx *context.APIContext) {
 
-	if splitted := strings.Split(ctx.Query("labels"), ","); len(splitted) > 0 {
+	splitted := strings.Split(ctx.Query("labels"), ",")
+	if len(splitted) > 0 {
```
[`len(splitted)` is always >= 1](https://play.golang.org/p/booAmz8TOtM). For example: `strings.Split("", ",") == []string{""}`, which has a `len()` of 1. You need to check whether `labels == ""` or `splitted[0] == ""` is empty too.

guillep2k replied:
What I mean is, you need to check `routers/api/v1/repo/issue.go:108` because the `if()` condition is wrong.

bhalbright replied:
Gotcha, now I get it (facepalm). I fixed the condition

bhalbright commented on routers/api/v1/repo/issue.go:
```diff
@@ -23,6 +24,131 @@ import (
+		}
+		log.Trace("Processing next %d repos of %d", len(repos), count)
+		for _, repo := range repos {
+			issueCount += repo.NumIssues
```
good catch, I fixed it

bhalbright commented on routers/api/v1/repo/issue.go:
```diff
@@ -23,6 +24,131 @@ import (
+	repoIDs := make([]int64, 0)
+	issueCount := 0
+	for page := 1; ; page++ {
+		repos, count, err := models.SearchRepositoryByName(&models.SearchRepoOptions{
```
thanks, I changed to return HTTP 400 if called when cross repo dependencies are disabled

guillep2k replied:
Sorry it took me so long to resume this review 😔. This PR looks good to me, except for this API function (`SearchIssues`) which handles too many options which I feel unrelated to the feature it's intended to serve. It should only search for the issues required for this PR, _especially_ if it will only work if `ALLOW_CROSS_REPOSITORY_DEPENDENCIES` is true. I understand the flexibility that parameters like `state` or `labels` can give, but they will not -currently- be used from the dependencies drop down box search. In fact, it should be named accordingly (e.g. `SearchIssuesForDependencies` or something).
Otherwise, I really like this feature. I've tested it with my repos and I think it's very useful and looks very nice. 🥇

bhalbright replied:
thanks @guillep2k! I guess what I was going for was since I needed a new api method to search issues across repos, I should follow the same pattern as the search function that already existed for searching one repo: "/repos/{owner}/{repo}/issues". I thought it would make sense to be feature parity with that endpoint. And then maybe it could be used by other use cases (like possibly this one: https://github.com/go-gitea/gitea/issues/2434). But I see your point, nothing else uses it at this point so it is debatable to make it general purpose. I can certainly strip it down if you still feel it should only match the functionality of the PR. Any additional thoughts?

zeripath replied:
I think the additional options make sense in that context - it would be annoying to have to add them later when the inevitable complaint comes in that the API is inconsistent. (We're probably going to have to expose all of the search repos options too though - which I'm not certain that this code does.)

guillep2k replied:
OK, then the check on `ALLOW_CROSS_REPOSITORY_DEPENDENCIES` to keep the client on adding them without permission should be moved to the function that adds the dependencies (where it should have been from the start, I think).

bhalbright replied:
@guillep2k I made the change to check ALLOW_CROSS_REPOSITORY_DEPENDENCIES when adding the dependency in issue_dependency.go instead. That's much better, and thanks again for the review.

bhalbright commented on custom/conf/app.ini.sample:
```diff
@@ -420,6 +420,8 @@ DEFAULT_ORG_MEMBER_VISIBLE = false
 ; Default value for EnableDependencies
 ; Repositories will use dependencies by default depending on this setting
 DEFAULT_ENABLE_DEPENDENCIES = true
+; The issue dependency search will show issues from all repositories where the user is granted access or only issues from the current repository depending on this setting.
```
sure, I made an adjustment

bhalbright commented on docs/content/doc/advanced/config-cheat-sheet.en-us.md:
```diff
@@ -246,6 +246,7 @@ Values containing `#` or `;` must be quoted using `` ` `` or `"""`.
 - `RECAPTCHA_SITEKEY`: **""**: Go to https://www.google.com/recaptcha/admin to get a sitekey for recaptcha.
 - `RECAPTCHA_URL`: **https://www.google.com/recaptcha/**: Set the recaptcha url - allows the use of recaptcha net.
 - `DEFAULT_ENABLE_DEPENDENCIES`: **true**: Enable this to have dependencies enabled by default.
+- `ALLOW_CROSS_REPOSITORY_DEPENDENCIES` : **true** Enable this to have the issue dependency search show issues from all repositories where the user is granted access. 
```
adjusted here too

6543 commented:
> should an issue that depends on a closed (unmerged) PR be considered unblocked?

I think it should be bloket, because you kan remove it from the PR later and this is intended then ...
And if a PR is bloket by something it normaly relys on this? So it make no senc if a dependency isnt merged so the blocket doesnt work ...

Didn't looked at the current behaviour but will do it later ...

bhalbright commented on routers/api/v1/repo/issue.go:
```diff
@@ -22,6 +23,137 @@ import (
+			TopicOnly:   false,
+			Collaborate: util.OptionalBoolNone,
+			UserIsAdmin: ctx.IsUserSiteAdmin(),
+			UserID:      ctx.Data["SignedUserID"].(int64),
```
Updated. You are right, it was a copy/paste error. Thanks for catching it.
